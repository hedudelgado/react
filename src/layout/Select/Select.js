import React from 'react';
import {
  string,
  bool,
  func,
  object,
  array,
} from 'prop-types';
import classnames from 'classnames';

const Select = ({
  placeholder,
  field,
  sessioncamhide,
  touched,
  error,
  id,
  label,
  children,
  flagImg,
  halfWidth,
  hideIcon,
  hideMessage,
  ...props
}) => {
  const showError = error && touched;
  const showValid = !error && touched;
  const selectClassNames = classnames({
    sessioncamhidetext: sessioncamhide,
    'error-field': showError,
  });
  const wrapperClassNames = classnames(
    'wb-form-item',
    { 'wb-push-middle--bottom': !halfWidth },
    { 'wb-native-dropdown-offset-flag': flagImg },
    { 'col mr2': halfWidth },
  );
  const childWrapperClassNames = classnames(
    'wb-form-item__field',
    { 'wb-form-item__field__full': !halfWidth },
    { 'wb-form-item__field__half': halfWidth },
  );
  const flagStyle = {};

  if (flagImg) {
    flagStyle.backgroundImage = `url(https://www.premierinn.com${flagImg})`;
  }

  return (
    <div className={wrapperClassNames}>
      {label && <label className="wb-strong" htmlFor={id}>{label}</label>}
      <div className={childWrapperClassNames}>
        <select
          className={selectClassNames}
          id={id}
          label={label}
          {...field}
          {...props}
        >
          {children}
        </select>
        <div className="wb-icon_arrow-down" />
        {flagImg && <span className="wb-native-dropdown--flag" style={flagStyle} />}
        {!hideIcon && showError && <span className="form-item-msg__error-icon wb-icon_disc-info-filled" />}
        {!hideIcon && showValid && <span className="form-item-msg__info-icon wb-icon_disc-checked-fill" />}
      </div>
      {!hideMessage && showError && <span className="form-item-msg form-item-msg--error">{error}</span>}
    </div>
  );
};

Select.propTypes = {
  /**
    HTML id attribute.
  */
  id: string.isRequired,
  /**
    HTML placeholder attribute.
  */
  placeholder: string,
  /**
    HTML value attribute.
  */
  value: string,
  /**
    Select label. Not displayed if empty.
  */
  label: string,
  /**
    Sessioncam field masking.
  */
  sessioncamhide: bool,
  /**
    Error message. Not displayed if empty.
  */
  error: string,
  /**
    Function to be run on select change.
  */
  onChange: func,
  field: object,
  touched: bool,
  children: array,
  flagImg: string,
  halfWidth: bool,
  hideIcon: bool,
  hideMessage: bool,
};

Select.defaultProps = {
  placeholder: '',
  halfWidth: false,
  hideIcon: false,
  hideMessage: false,
};

export default Select;
