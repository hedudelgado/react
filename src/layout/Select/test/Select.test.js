import React from 'react';
import { shallow } from 'enzyme';

import Select from '../Select';

describe('Select Component', () => {
  const props = {
    id: 'test',
    type: 'text',
    placeholder: 'Placeholder',
    value: 'value',
    label: 'Label',
    touched: true,
  };

  it('should render the component', () => {
    const wrapper = shallow(<Select {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render label if not set', () => {
    const wrapper = shallow(<Select {...{ ...props, label: null }} />);
    expect(wrapper).not.toContainMatchingElement('label');
  });

  it('should render the error if set', () => {
    const wrapper = shallow(<Select {...{ ...props, error: 'error message' }} />);
    expect(wrapper).toContainMatchingElement('.error-field');
    expect(wrapper).toContainMatchingElement('.form-item-msg--error');
    expect(wrapper.find('.form-item-msg--error')).toHaveText('error message');
  });

  it('should set sessioncamhide class if set', () => {
    const wrapper = shallow(<Select {...{ ...props, sessioncamhide: true }} />);
    expect(wrapper.find('select')).toHaveClassName('sessioncamhidetext');
  });

  it('should show flagImg when variable is set', () => {
    const wrapper = shallow(<Select {...{ ...props, sessioncamhide: true, flagImg: 'flagImg' }} />);
    expect(wrapper.find('.wb-native-dropdown--flag')).toHaveLength(1);
  });
});
