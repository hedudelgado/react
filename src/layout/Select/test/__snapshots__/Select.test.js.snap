// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component should render the component 1`] = `
<div
  className="wb-form-item wb-push-middle--bottom"
>
  <label
    className="wb-strong"
    htmlFor="test"
  >
    Label
  </label>
  <div
    className="wb-form-item__field wb-form-item__field__full"
  >
    <select
      className=""
      id="test"
      label="Label"
      type="text"
      value="value"
    />
    <div
      className="wb-icon_arrow-down"
    />
    <span
      className="form-item-msg__info-icon wb-icon_disc-checked-fill"
    />
  </div>
</div>
`;
